dynamiqs/solvers/propagator.py

```diff
@@ -1,12 +1,24 @@
+from __future__ import annotations
+
 from abc import abstractmethod
 
+import numpy as np
 from torch import Tensor
 
 from .solver import AutogradSolver
 from .utils.td_tensor import ConstantTDTensor
 from .utils.utils import tqdm
 
 
+def round_truncate(x: np.float32 | np.float64) -> np.float32 | np.float64:
+    # round a strictly positive-valued float to remove numerical errors, and enable
+    # comparing floats for equality
+
+    # round to 5 additional decimal places beyond the highest digit of the number
+    decimals = abs(int(np.log10(x))) + 5
```

Comment:
I think it would make more sense to do this in binary (also easier to adapt between float16/float32/float64). So `decimals = abs(int(np.log2(x))) + 16` maybe? (there are 23 bits allocated to the fraction in a float32, and 2**16 = 65536, `20` could be good also)

Reply:
Why not, but for some reason the truncation is less satisfactory:

For decimals:
```
0.00009999999747378751635551452637 -> 0.00010000000000000000479217360239
0.00009999999747378751635551452637 -> 0.00010000000000000000479217360239
0.00010000001930166035890579223633 -> 0.00010000000000000000479217360239
0.00009999997564591467380523681641 -> 0.00010000000000000000479217360239
0.00010000003385357558727264404297 -> 0.00010000000000000000479217360239
0.00010000000474974513053894042969 -> 0.00010000000000000000479217360239
0.00009999994654208421707153320312 -> 0.00010000000000000000479217360239
0.00010000000474974513053894042969 -> 0.00010000000000000000479217360239
0.00010000000474974513053894042969 -> 0.00010000000000000000479217360239
0.00010000006295740604400634765625 -> 0.00010000000000000000479217360239
```

For binary:
```text
0.00009999999747378751635551452637 -> 0.00009999983012676239013671875000
0.00009999999747378751635551452637 -> 0.00009999983012676239013671875000
0.00010000001930166035890579223633 -> 0.00009999983012676239013671875000
0.00009999997564591467380523681641 -> 0.00009999983012676239013671875000
0.00010000003385357558727264404297 -> 0.00009999983012676239013671875000
0.00010000000474974513053894042969 -> 0.00009999983012676239013671875000
0.00009999994654208421707153320312 -> 0.00009999983012676239013671875000
0.00010000000474974513053894042969 -> 0.00009999983012676239013671875000
0.00010000000474974513053894042969 -> 0.00009999983012676239013671875000
0.00010000006295740604400634765625 -> 0.00009999983012676239013671875000
```

I'm not sure it makes a huge difference, I also made a distinction between different types of float but then thought that it's a completely extreme use case to need such high precision on the `delta_t`.

Reply:
As you wish, I prefer the binary version, but it's not super important in any case. The two rounded numbers you're showing are probably the same in float32 precision, and it's kinda random whether you're closer or not to the wanted `delta_t`.

Reply:
Alright, I tried to make the comment as clear as possible.